client/rest/src/main/java/org/elasticsearch/client/LanguageRuntimeVersions.java

```diff
@@ -57,79 +54,74 @@ public static String getRuntimeMetadata() {
     }
 
-    public static String kotlinVersion() {
+    public static String scalaVersion() {
+        // scala.util.Properties.versionNumberString()
+        return keepMajorMinor(callStaticMethod("scala.util.Properties", "versionNumberString"));
+    }
+
+    public static String clojureVersion() {
+        // (clojure-version) which translates to
+        // clojure.core$clojure_version.invokeStatic()
+        return keepMajorMinor(callStaticMethod("clojure.core$clojure_version", "invokeStatic"));
+    }
+
+    public static String groovyVersion() {
+        // groovy.lang.GroovySystem.getVersion()
+        // There's also getShortVersion(), but only since Groovy 3.0.1
+        return keepMajorMinor(callStaticMethod("groovy.lang.GroovySystem", "getVersion"));
+    }
+
+    public static String jRubyVersion() {
+        // org.jruby.runtime.Constants.VERSION
+        return keepMajorMinor(getStaticField("org.jruby.runtime.Constants", "VERSION"));
+    }
+
+    private static String getStaticField(String className, String fieldName) {
+        Class<?> clazz;
         try {
-            //KotlinVersion.CURRENT.toString()
-            Class<?> clazz = Class.forName("kotlin.KotlinVersion");
-            Field field = clazz.getField("CURRENT");
-            String version = field.get(null).toString();
-            return stripPatchRevision(version);
-
-        } catch (Exception t) {
-            // ignore
+            clazz = Class.forName(className);
+        } catch (ClassNotFoundException e) {
+            return null;
         }
-        return null;
-    }
 
-    public static String scalaVersion() {
         try {
-            // scala.util.Properties.versionNumberString()
-            Class<?> clazz = Class.forName("scala.util.Properties");
-            Method m = clazz.getMethod("versionNumberString");
-            String version = (String) m.invoke(null);
-            return stripPatchRevision(version);
-
-        } catch (Exception t) {
-            // ignore
+            Field field = clazz.getField(fieldName);
+            return field.get(null).toString();
+        } catch (Exception e) {
+            return ""; // can't get version information
         }
-        return null;
     }
 
-    public static String clojureVersion() {
+    private static String callStaticMethod(String className, String methodName) {
+        Class<?> clazz;
         try {
-            // (clojure-version) which translates to
-            // clojure.core$clojure_version.invokeStatic()
-            Class<?> clazz = Class.forName("clojure.core$clojure_version");
-            Method m = clazz.getMethod("invokeStatic");
-            String version = (String) m.invoke(null);
-            return stripPatchRevision(version);
-
-        } catch (Exception t) {
-            // ignore
+            clazz = Class.forName(className);
+        } catch (ClassNotFoundException e) {
+            return null;
```

Comment:
Is the `return null` vs `return ""` difference purposeful? Isn't it the same "can't get the version" case?

Reply:
Yes, it's slightly different: we return `null` when the class doesn't exist, i.e. we're sure the language runtime isn't there (or at least the class identifying it). When it exists and we can't call the method or can't get the field, this means the information to look for has changed, and we report the language as existing but without version information. That will trigger an analysis on our end to look for updates to the language runtime.

Reply:
Hm, but if `org.jruby.runtime.Constants` is moved to `org.jruby.newpackage.Constants` this invariant will not hold anymore (ie. a particular class being there or not finding for the correct class FQN are indistinguishable cases IMO). I don't have a strong opinion on this, but I'd personally go for the simpler option of "I don't know the version or if it is even present at all"